### PR TITLE
MM-42542 Set up @mattermost/types for publishing to NPM

### PR DESCRIPTION
**Pull request by hmhealey**
Trying to ship these as part of the client without making them a separate package kind of caused a bunch of problems, so instead we're going to ship them separately, and everyone who wants to use them will need both.

Some implementation details:

- I talk about it more in `packages/types/README.md`, but these types will be importable using subpaths like before (eg. `import {UserProfile} from '@mattermost/types/users'`) because of the sheer number of things we're exporting from this package. This uses the `exports` field in `package.json` which is supported in some tools, but not yet in others, so I included steps to support this in those tools.

    I also considered going with how mattermost-redux used to be shipped where everything was compiled into the root of the package, and while that would work everywhere, this seemed a bit nicer on our end since you won't see a load of ignored, compiled files in your editor.
- I ended up going with just `tsc` to compile these type definitions since they're fairly simple and self-contained. I could see changing to a different build tool for this once we find one that works for other packages though.
- Finally, instead of having to build this live, I just set up the web app to import the source for these directly. That seems easier for the time being even if it does have the potential for them to work differently in the web app than in other projects.

Next steps:
- Do the same for `Client4` with `@mattermost/client`. That'll let me see which types are actually needed for that and which can stay in `mattermost-redux`.
- Remove the internal types from this package. I went with moving everything first and then moving stuff back to the web app after since that seemed less likely to cause conflicts than if I modified each type file on its own.
- Similarly, all the code in the web app still references these types via mattermost-redux by having it re-export the types from the new package. Once I've determined which types go where, then I'll go update references.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-42542

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by hmhealey** [comment]
Note that I'll be force pushing this branch to resolve conflicts. Any change to the type files will cause a conflict because I'm moving them, and repeatedly rebasing is by far the easiest way to resolve that. That's also why I moved the files in one commit before making any changes to them in the followup

File: packages/types/README.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        "paths": {
+            "@mattermost/types/*": ["node_modules/@mattermost/types/lib/*"]
+        }
+    }
+}
+```
+
+### Jest
+
+In your Jest config, you can use the `moduleNameMapper` field to adsd that alias.
```
Comment (M-ZubairAhmed) [inline packages/types/README.md:34]:
```suggestion
In your Jest config, you can use the `moduleNameMapper` field to add that alias.
```

**Review by M-ZubairAhmed — APPROVED** [review]
This is a really good effort. Also do we plan on eventually removing the types from mm-redux, so we dont have to export * from each file. This effort can be a good campaign i guess

**Comment by hmhealey** [comment]
Yeah, that's going to come a bit later. Next step is to ship the client package using these types, then I'll move the internal types not used by it back into their old location, and finally I'll update imports throughout the rest of the app.

**Review by crspeller — APPROVED** [review]
Nice! Good first steps to splitting these out.

**Pull request merged**